gr-uhd/lib/rfnoc_replay_impl.cc

```diff
@@ -0,0 +1,201 @@
+    }
+
+    // Make sure, we use dicts!
+    if (!pmt::is_dict(msg)) {
+        d_logger->error("Command message is neither dict nor pair: {}",
+                        pmt::write_string(msg));
+        return;
+    }
+
+    if (!pmt::dict_has_key(msg, replay_cmd_key())) {
+        d_logger->error("Command message does not have a command: {}",
+                        pmt::write_string(msg));
+        return;
+    }
+
+    bool has_cmd_offset = false;
+    bool has_cmd_size = false;
+    uint64_t cmd_offset, cmd_size;
+
+    // Check to see if the command sent specifies a size and/or offset for the command
+    // If it does not that is fine, we will use the record/play size/offset properties.
+    if (pmt::dict_has_key(msg, replay_cmd_offset_key())) {
+        cmd_offset = int(
+            pmt::to_long(pmt::dict_ref(msg, replay_cmd_offset_key(), pmt::from_long(0))));
+        has_cmd_offset = true;
+    }
+    if (pmt::dict_has_key(msg, replay_cmd_size_key())) {
+        cmd_size = int(
+            pmt::to_long(pmt::dict_ref(msg, replay_cmd_size_key(), pmt::from_long(0))));
+        has_cmd_size = true;
+    }
+
+    int cmd_port = int(pmt::to_long(pmt::dict_ref(msg,
+                                                  replay_cmd_port_key(),
+                                                  pmt::from_long(0) // Default to Port 0
+                                                  )));
+
+    double cmd_time =
+        int(pmt::to_float(pmt::dict_ref(msg,
+                                        replay_cmd_time_key(),
+                                        pmt::from_float(0.0) // Default to 0.0
+                                        )));
+
+    std::string command =
+        pmt::symbol_to_string(pmt::dict_ref(msg, replay_cmd_key(), pmt::intern("")));
+
+    if (command == "record") {
+        // If command didn't specify size/offset get from the record properties
+        if (!has_cmd_offset)
+            cmd_offset = d_wrapped_ref->get_record_offset(cmd_port);
+        if (!has_cmd_size)
+            cmd_size = d_wrapped_ref->get_record_size(cmd_port);
+        record(cmd_offset, cmd_size, cmd_port);
+        return;
+    }
+    // If command didn't specify size/offset get from the play properties
+    if (!has_cmd_offset)
+        cmd_offset = d_wrapped_ref->get_play_offset(cmd_port);
+    if (!has_cmd_size)
+        cmd_size = d_wrapped_ref->get_play_size(cmd_port);
+    if (command == "play") {
+        play(cmd_offset, cmd_size, cmd_port, cmd_time, false);
+        return;
+    }
+    if (command == "play_continuous") {
+        play(cmd_offset, cmd_size, cmd_port, cmd_time, true);
+        return;
+    }
+    if (command == "stop") {
+        stop();
+        return;
+    }
+    if (command == "get_record_fullness") {
+        d_logger->debug("Record fullness of port {:d}: {:d}",
+                        cmd_port,
+                        d_wrapped_ref->get_record_fullness(cmd_port));
```

Comment:
This is a nice addition -- but no good deed (or good idea) goes unpunished. This could be even more useful if you post the fullness on an output message port instead of logging it. For those who want to log it, no problem, they can simply connect the output message port to a message logger. But this would allow other blocks to programmatically query the fullness, and maybe act upon that.